### DISCUSSION TITLE
Environmental Variable Issue

### DISCUSSION THREAD

**Discussion by VikasAgrawal-04** [codemagic-ci-cd/codemagic-docs#2671] [Q&A] [answered]
### If you are looking for help, please confirm the following...

- [X] I confirm I have searched the [Docs](https://docs.codemagic.io), [Codemagic Sample Projects](https://github.com/codemagic-ci-cd/codemagic-sample-projects), and GitHub [Discussions](https://github.com/orgs/codemagic-ci-cd/discussions).
- [X] I confirm the software versions on my local computer match the ones in CI environment (e.g. Xcode version, Gradle version). See [here](https://docs.codemagic.io/troubleshooting/common-android-issues/#builds-work-locally-but-fail-on-codemagic)
- [X] I confirm I am able to do a fresh clone of the repository and build the project on my local computer.

### Which mobile framework are you using?

Flutter (Dart)

### Steps to reproduce

1. Adding Environmental Vairable.
2. Using those variable in code with $Key
![Screenshot from 2024-08-07 14-34-50](https://github.com/user-attachments/assets/60b210f6-cea3-4093-8c4f-ef26f4d933e4)

### Expected results

It should get the Key value from the Codemagic env variables.

### Actual results

It is not able to load the variables
![Screenshot from 2024-08-07 14-36-52](https://github.com/user-attachments/assets/e9847b67-eded-4d50-9eda-8ab18ebcace6)

**Comment by rshrc** [comment]
Facing the same issue!! environment variables are a pain, confused!! ;-;

**Comment by dam-ease** [comment, marked as answer]
Hello @VikasAgrawal-04.
I see you are trying to access the Environment Variables in your codebase. 

One way to do this is:

- Add your environment variables to Codemagic UI as you have done in screenshots you shared
- When building the flutter app, add `--dart-define` to your scripts. E.g `--dart-define=NAME=$NAME`
<img width="633" alt="Screenshot 2024-08-07 at 11 12 49" src="https://github.com/user-attachments/assets/4c8194e5-cee2-484f-b1e6-751c02c41c12">

- You can now access these variables in your code with `String.fromEnvironment` method. Example: `String.fromEnvironment('NAME')`

You can also checkout [this](https://www.sandromaglione.com/articles/how-to-use-environmental-variables-in-flutter) for more ways of handling environment variables. Please let me know if this helps.

**Reply by dam-ease** [reply]
Another easy way to go about this is by creating a local .env file in a pre-build script and then populate the variables to .env during the build:

```
echo "API_KEY=$API_KEY" > .env
echo "CLIENT_SECRET=$CLIENT_SECRET" >> .env
```
